File: common/src/main/java/com.abc/model/rest/UserResponse.java
package com.abc.model.rest;

import java.util.UUID;

public abstract class UserResponse {

    private UUID id;

    public UserResponse(UUID id) {
        this.id = id;
    }

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }
}

File: common/src/main/java/com.abc/adapter/UserResponse.kt
package com.abc.adapter

import com.abc.model.rest.DecisionResponse
import com.google.gson.*
import java.lang.reflect.Type

class UserResponse : JsonSerializer<DecisionResponse>, JsonDeserializer<DecisionResponse> {

    override fun serialize(src: DecisionResponse, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val res = JsonObject();
        res.add("type", JsonPrimitive(src.javaClass.simpleName))
        res.add("props", context.serialize(src, src.javaClass))
        return res
    }

    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): DecisionResponse {
        val jsonObj = json.asJsonObject
        val type = jsonObj["type"].asString
        val element = jsonObj["props"]
        return context.deserialize(element, Class.forName("com.abc.model.rest.$type"))
    }

}